File: frontend/src/utils/validaciones.js
import { isValidPhoneNumber } from 'react-phone-number-input';

export function validarTelefono(valor) {
  if (!valor) return null; // opcional
  if (!isValidPhoneNumber(valor)) return "Número de teléfono internacional inválido.";
  return null;
}

export function validarEmail(valor) {
  if (!valor) return null; // opcional
  const re = /^[^\s@]+@[^\s@]+\.[^\s@]{2,}$/;
  if (!re.test(valor)) return "Correo electrónico inválido.";
  return null;
}

export function normalizarCedula(valor) {
  if (!valor) return valor;
  const limpio = valor.trim().replace(/\s/g, "");
  if (/^\d{6,8}$/.test(limpio)) return `V-${limpio}`;
  return limpio;
}

export function validarCedula(valor) {
  if (!valor || !valor.trim()) return null; // opcional
  const limpio = normalizarCedula(valor);
  if (!/^[VvEe]-?\d{6,8}$/.test(limpio))
    return "Debe iniciar con V o E. Ej: V-12345678.";
  return null;
}

export function validarFormMiembro(f) {
  const errores = {};
  if (!f.nombre.trim()) errores.nombre = "El nombre es obligatorio.";

  const eCed = validarCedula(f.cedula);
  if (eCed) errores.cedula = eCed;

  if (!f.telefono.trim())         errores.telefono = "El teléfono principal es obligatorio.";
  else { const e = validarTelefono(f.telefono); if (e) errores.telefono = e; }

  if (f.tlf_alternativo?.trim()) { const e = validarTelefono(f.tlf_alternativo); if (e) errores.tlf_alternativo = e; }

  if (!f.email.trim())            errores.email = "El correo es obligatorio.";
  else { const e = validarEmail(f.email); if (e) errores.email = e; }

  if (!f.cargo) errores.cargo = "El cargo es obligatorio.";

  return errores;
}

File: backend/app/validaciones.py
import re
import phonenumbers

CARGOS = {"Profesor","Estudiante","BR","Auxiliar","Voluntario"}

def validar_telefono(valor):
    if not valor:
        return None
        
    # Excepción: Permitir el nuevo código 0422 de Digitel (Venezuela) que puede no estar en la librería aún
    limpio = valor.replace(" ", "")
    if limpio.startswith("+58422") and len(limpio) == 13:
        return None
        
    try:
        parsed = phonenumbers.parse(valor, None)
        if not phonenumbers.is_valid_number(parsed):
            return "Número de teléfono internacional inválido."
    except phonenumbers.NumberParseException:
        return "Formato inválido. Asegúrese de incluir el código de país (ej. +58)."
    return None

def validar_email(valor):
    if not valor:
        return None
    if not re.fullmatch(r"[^\s@]+@[^\s@]+\.[^\s@]{2,}", valor):
        return "Correo electrónico inválido."
    return None

def normalizar_cedula(valor):
    """Si solo tiene dígitos, agrega V- por defecto."""
    if not valor:
        return valor
    limpio = valor.strip().replace(" ", "")
    if re.fullmatch(r"\d{6,8}", limpio):
        return f"V-{limpio}"
    return limpio

def validar_cedula(valor):
    if not valor or not valor.strip():
        return "La cédula es obligatoria."
    limpio = normalizar_cedula(valor)
    if not re.fullmatch(r"[VvEe]-?\d{6,8}", limpio):
        return "Debe iniciar con V o E. Ej: V-12345678."
    return None

def _s(val):
    """Safely strip a value that may be None."""
    return (val or "").strip()

def validar_miembro(data):
    errores = {}
    if not _s(data.get("nombre")):
        errores["nombre"] = "El nombre es obligatorio."
    if _s(data.get("cedula")):
        e = validar_cedula(data.get("cedula"))
        if e: errores["cedula"] = e
    if not _s(data.get("telefono")):
        errores["telefono"] = "El teléfono principal es obligatorio."
    else:
        e = validar_telefono(data.get("telefono"))
        if e: errores["telefono"] = e
    if _s(data.get("tlf_alternativo")):
        e = validar_telefono(data.get("tlf_alternativo"))
        if e: errores["tlf_alternativo"] = e
    if not _s(data.get("email")):
        errores["email"] = "El correo es obligatorio."
    else:
        e = validar_email(data.get("email"))
        if e: errores["email"] = e
    if not _s(data.get("cargo")):
        errores["cargo"] = "El cargo es obligatorio."
    elif data.get("cargo") not in CARGOS:
        errores["cargo"] = f"Cargo inválido. Opciones: {', '.join(CARGOS)}."
    return errores
